Assets/AWSIM/Scripts/Vehicles/StuckSensor.cs

```diff
@@ -0,0 +1,153 @@
+        {
+            if (vehicleCurrentSpeed > maxSpeedWhileStuck || Math.Abs(vehicleTargetSpeed) < 0.01)
+                distanceExceededCounter = 0;
+            traveledDistance += vehicleCurrentSpeed * Time.deltaTime;
+            targetDistance += vehicleTargetSpeed * Time.deltaTime;
+            var isThresholdExceeded = targetDistance - traveledDistance > distanceDifferenceThreshold;
+            var isEndOfDistanceSummation = (System.DateTime.Now - startDistanceSumTimestamp).TotalSeconds > distanceTimePeriod;
+            if (isEndOfDistanceSummation)
+            {
+                traveledDistance = 0.0;
+                targetDistance = 0.0;
+                startDistanceSumTimestamp = System.DateTime.Now;
+                if (isThresholdExceeded)
+                    distanceExceededCounter++;
+                else
+                    distanceExceededCounter--;
+                if (distanceExceededCounter < 0) distanceExceededCounter = 0;
+                if (distanceExceededCounter > maxDistanceExceededCounterValue) distanceExceededCounter = maxDistanceExceededCounterValue;
+            }
+            return distanceExceededCounter > maxDistanceExceededCounterValue / 2;
+        }
+
+        void FixedUpdate()
+        {
+            if (initialized == false)
+                return;
+
+            if (Math.Abs(vehicleCurrentSpeed) > 0.01 || Math.Abs(vehicleTargetSpeed) < 0.01)
+                lastMoveTimestamp = System.DateTime.Now;
+
+            //case when the collision occurs constantly
+            var isStuckCollision = false;
+            if (considerCollisionStay)
+                isStuckCollision = (System.DateTime.Now - lastCollisionTimestamp).TotalSeconds < collisionTimeThreshold;
+            
+            //case when vehicle should be moving (receives speed) but vehicle isn't moving as it should
+            var isDistanceDifferenceExceeded = false;
+            if (considerDistanceTraveledOverTime)
+                isDistanceDifferenceExceeded = IsDistanceDifferenceExceeded();
+
+            //case when vehicle should be moving but the distanceDifferenceThreshold is not exceeded
+            //becouse vehicleTargetSpeed is very small and vehicleCurrentSpeed==0 for a long time
+            var isLongImmobilityExceeded = false;
+            if (considerLongImmobility)
+            {
+                var longImmobility = Math.Abs(vehicleTargetSpeed) > 0.01 && (System.DateTime.Now - lastMoveTimestamp).TotalSeconds > longImmobilityTimeThreshold;
+                if(longImmobility) lastImmobilityTimestamp = System.DateTime.Now;
+                isLongImmobilityExceeded = (System.DateTime.Now - lastImmobilityTimestamp).TotalSeconds < longImmobilityTimeThreshold;
+            }
+
+            stuckMsg.Data = isStuckCollision || isDistanceDifferenceExceeded || isLongImmobilityExceeded;
+            if (!NeedToPublish())
+                stuckPublisher.Publish(stuckMsg);
+        }
+
+        void OnGUI()
```

Comment:
As for the UI, it may be moved to the uGUI canvas in the future. May I ask you to mention this in your comments?

Also, it would be nice to have a parameter in the inspector to turn the UI display on and off with a bool.

Reply:
I've provided ` on/off ` information in the GUI - I also added the mentioned comment.